File: src/main/java/com/example/chatService/redis/UserSessionRegistry.java
package com.example.chatService.redis;

import com.example.chatService.dto.UserRoomSession;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
@RequiredArgsConstructor
public class UserSessionRegistry {

    private static final String PREFIX = "ws:session:";
    private static final Duration TTL = Duration.ofHours(6);

    private final StringRedisTemplate redisTemplate;
    private final ObjectMapper objectMapper;

    public void save(String sessionId, String userId, String roomId) {
        UserRoomSession value = new UserRoomSession(userId, roomId);

        try {
            String json = objectMapper.writeValueAsString(value);
            redisTemplate.opsForValue().set(key(sessionId), json, TTL);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("웹소켓 세션 저장 실패", e);
        }
    }

    public UserRoomSession get(String sessionId) {
        String json = redisTemplate.opsForValue().get(key(sessionId));
        if (json == null) {
            return null;
        }

        try {
            return objectMapper.readValue(json, UserRoomSession.class);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("웹소켓 세션 조회 실패", e);
        }
    }

    public UserRoomSession remove(String sessionId) {
        String redisKey = key(sessionId);
        String json = redisTemplate.opsForValue().get(redisKey);

        if (json == null) {
            return null;
        }

        redisTemplate.delete(redisKey);

        try {
            return objectMapper.readValue(json, UserRoomSession.class);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("웹소켓 세션 삭제 실패", e);
        }
    }

    private String key(String sessionId) {
        return PREFIX + sessionId;
    }
}

File: src/main/java/com/example/chatService/redis/UserSessionRegistry.kt
package com.example.chatService.redis

import com.example.chatService.dto.UserRoomSession
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class UserSessionRegistry(
    private val redisTemplate: StringRedisTemplate,
    private val objectMapper: ObjectMapper,
) {
    fun save(sessionId: String, userId: String, roomId: String) {
        val value = UserRoomSession(userId, roomId)
        try {
            val json = objectMapper.writeValueAsString(value)
            redisTemplate.opsForValue().set(key(sessionId), json, TTL)
        } catch (e: JsonProcessingException) {
            throw IllegalStateException("웹소켓 세션 저장 실패", e)
        }
    }

    fun get(sessionId: String): UserRoomSession? {
        val json = redisTemplate.opsForValue().get(key(sessionId)) ?: return null
        try {
            return objectMapper.readValue(json, UserRoomSession::class.java)
        } catch (e: JsonProcessingException) {
            throw IllegalStateException("웹소켓 세션 조회 실패", e)
        }
    }

    fun remove(sessionId: String): UserRoomSession? {
        val redisKey = key(sessionId)
        val json = redisTemplate.opsForValue().get(redisKey) ?: return null
        redisTemplate.delete(redisKey)

        try {
            return objectMapper.readValue(json, UserRoomSession::class.java)
        } catch (e: JsonProcessingException) {
            throw IllegalStateException("웹소켓 세션 삭제 실패", e)
        }
    }

    private fun key(sessionId: String): String = "$PREFIX$sessionId"

    companion object {
        private const val PREFIX = "ws:session:"
        private val TTL: Duration = Duration.ofHours(6)
    }
}
